x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java

```diff
@@ -0,0 +1,116 @@
+import org.elasticsearch.xpack.shutdown.DeleteShutdownNodeAction;
+import org.elasticsearch.xpack.shutdown.PutShutdownNodeAction;
+import org.elasticsearch.xpack.shutdown.ShutdownPlugin;
+import org.hamcrest.Matchers;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, minNumDataNodes = 2)
+public class SearchableSnapshotShutdownIntegTests extends BaseSearchableSnapshotsIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.appendToCopy(super.nodePlugins(), ShutdownPlugin.class);
+    }
+
+    public void testAllocationDisabledDuringShutdown() throws Exception {
+        final String restoredIndexName = setupMountedIndex();
+        final Set<String> indexNodes = internalCluster().nodesInclude(restoredIndexName);
+        final ClusterState state = client().admin().cluster().prepareState().clear().setRoutingTable(true).setNodes(true).get().getState();
+        final Map<String, String> nodeNameToId = state.getNodes()
+            .stream()
+            .collect(Collectors.toMap(DiscoveryNode::getName, DiscoveryNode::getId));
+
+        for (String indexNode : indexNodes) {
+            final String indexNodeId = nodeNameToId.get(indexNode);
+            putShutdown(indexNodeId);
+            final int shards = (int) state.routingTable()
+                .allShards(restoredIndexName)
+                .stream()
+                .filter(s -> indexNodeId.equals(s.currentNodeId()))
+                .count();
+            assert shards > 0;
+
+            final CacheService cacheService = internalCluster().getInstance(CacheService.class, indexNode);
+            cacheService.synchronizeCache();
+
+            logger.info("--> Restarting [{}/{}]", indexNodeId, indexNode);
+            internalCluster().restartNode(indexNode, new InternalTestCluster.RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String nodeName) throws Exception {
+                    assertBusy(() -> {
+                        ClusterHealthResponse response = client().admin()
+                            .cluster()
+                            .health(Requests.clusterHealthRequest(restoredIndexName))
+                            .actionGet();
+                        assertThat(response.getUnassignedShards(), Matchers.equalTo(shards));
+                    });
+                    return super.onNodeStopped(nodeName);
+                }
+            });
+            // leave shutdown in place for some nodes to check that shards get assigned anyway.
+            if (randomBoolean()) {
```

Comment:
I wonder if this can cause some test failures when a low number of data nodes is picked up? I quickly tested with a 10 shards index and 2 data nodes and got a failure after ~5 runs.